### PR TITLE
[[ DataGrid 2 ]] Edit mode and swipe actions.

### PR DESCRIPTION
**Pull request by livecodemichael**
The DataGrid has been extended to include a new edit mode and swipe actions.

The following properties have been added to the DataGrid:
- set the dgEditMode of group "dg" to <true|false>
- set the dgProps["animate actions"] of group "dg" to <true|false>
- set the dgProps["enable swipe"] of group "dg" to <true|false>
- set the dgProps["edit mode action control"] of group "dg" to \<control id\>
- set the dgProps["edit mode action select control"] of group "dg" to \<control id\>
- set the dgProps["edit mode reorder control"] of group "dg" to \<control id\>
- set the dgProps["left swipe control"] of group "dg" to \<control id\>
- set the dgProps["right swipe control"] of group "dg" to \<control id\>

The following commands have been added to the DataGrid:
- dispatch "EditModeShowActionControlForIndex" to group "dg" with \<index\>
- dispatch "EditModeShowActionControlForIndex" to group "dg" with <true|false>
- dispatch "RowSwipeShowControlForIndexAndSide" to group "dg" with \<index\>, \<left|right\>
- dispatch "RowSwipeHideControl" to group "dg" with <true|false>

The following messages are sent to the DataGrid:
- on EditModeReorderStarted pIndex, pLineNo
- on EditModeReorderCompleted pIndex, pStartLineNo, pNewLineNo

The following messages are sent to the appropriate row control of the DataGrid:
- on EditModeActionSelectControlClicked pTarget
- on EditModeActionControlClicked pTarget
- on EditModeActionControlHidden
- on RowSwipedRight
- on RowSwipedLeft
- on RowLeftSwipeControlClicked pTarget
- on RowRightSwipeControlClicked pTarget
- on RowLeftSwipeControlHidden
- on RowRightSwipeControlHidden

The LayoutControl message sent to the row control has been updated to add a second parameter:
- on LayoutControl pControlRect, pWorkingRect

The following messages are sent to the each row control of the DataGrid, allowing for control customisation:
- on GetEditModeActionControl
- on GetEditModeActionSelectControl
- on GetEditModeReorderControl
- on GetLeftSwipeControl
- on GetRightSwipeControl

revdatagridlibrary.rev has been updated to include the new behaviours and an updated template row behaviour script.

### CODE REVIEW

File: Documentation/dictionary/datagrid.lcdoc [hunk 1]
```diff
@@ -2613,3 +2615,918 @@ Description:
...
+
+
+Name: RowSwipeShowControlForIndexAndSide
+
+Type: command
+
+Associations: datagrid, datagrid commands
+
+Syntax: dispatch "RowSwipeShowControlForIndexAndSide" to group "DataGrid" with
+<row index>, {left | right}
```
Comment (livecodeali) [inline Documentation/dictionary/datagrid.lcdoc]: I think `"left"` and `"right"` should be in quotes here

File: Documentation/dictionary/datagrid.lcdoc [hunk 2]
```diff
@@ -2613,3 +2615,918 @@ Description:
...
+end EditModeActionSelectControlClicked
+
+Description:
+
+<EditModeActionSelectControlClicked> is sent to your data grid's custom row
+template when the user clicks on a row's action select control. Handle
+<EditModeActionSelectControlClicked> if you wish to perform a custom behavior.
+
+pTarget is the target of the click. Use this to determine where in the action
+control the user clicked.
```
Comment (livecodeali) [inline Documentation/dictionary/datagrid.lcdoc]: You can put the info about `pTarget` in a 
```
Parameters:
pTarget: 
```
section

File: Documentation/dictionary/datagrid.lcdoc [hunk 3]
```diff
@@ -2613,3 +2615,918 @@ Description:
...
+end EditModeActionControlClicked
+
+Description:
+
+<EditModeActionControlClicked> is sent to your data grid's custom row template
+when the user clicks on a row's action control. Handle
+<EditModeActionControlClicked> if you wish to perform a custom behavior.
+
+pTarget is the target of the click. Use this to determine where in the action
+control the user clicked.
```
Comment (livecodeali) [inline Documentation/dictionary/datagrid.lcdoc]: And again here

File: Documentation/dictionary/datagrid.lcdoc [hunk 4]
```diff
@@ -2613,3 +2615,918 @@ Description:
...
+Summary: Sent when a user completes a dynamic reordering of a data grid.
+
+Description:
+
+<EditModeReorderCompleted> is sent to your data grid when a user completes a
+dynamic reordering.
+
+<pIndex> is the index of the reordered row. <pStartLineNo> is the original line
+sequence number of the row within the data grid. <pNewLineNo> is the new line
+sequence number of the row within the data grid.
```
Comment (livecodeali) [inline Documentation/dictionary/datagrid.lcdoc]: These should be documented as parameters

File: Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript [hunk 5]
```diff
@@ -68,6 +68,112 @@ local sTableObjectsA
...
+local sEditMode
+
+---
+
+local sReorderInProgress
+local sReorderControl
+local sReorderStartIndex
+local sReorderStartLine
+local sReorderLastHoverLine
+local sReoderMouseMoveInProgress
```
Comment (livecodeali) [inline Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript]: Not a bug as you've been consistent, but there's a typo in this variable name

File: Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript [hunk 6]
```diff
@@ -68,6 +68,112 @@ local sTableObjectsA
...
+constant kReorderDragControlGlowColor = "0,0,0"
+constant kReorderDragControlGlowSize = 10
+constant kReorderDragControlGlowSpread = 10
+
+---
+
+local sReorderScrollPollMsgID
+
+constant kReorderScrollPollRate = 30
+constant kReoderScrollPollMargin = 30
```
Comment (livecodeali) [inline Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript]: Another `reoder` here

File: Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript [hunk 7]
```diff
@@ -68,6 +68,112 @@ local sTableObjectsA
...
+
+constant kSwipeReturnDuration = 500
+constant kSwipeReturnEasing = "out bounce"
+
+---
+
+local sCustomisableControlsInitialised
+local sCustomisableControlsA
+local sConstomisableControlsSingletonsA
+local sConstomisableControlsSingletonCopyMapA
```
Comment (livecodeali) [inline Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript]: Some `constomisable` rather than `customisable`s here

**Comment by livecodeali** [comment]
This looks great as far as I can tell, other than a couple of docs things and non-critical typos. One minor suggestion I have is that the `Get...Control` messages be `function`s rather than `on` - I see you are constructing the dispatch message generically, but am I right in thinking the ones that are expected to return something are done separately (in which case there's no reason not to use `dispatch function`)?

File: Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript (hunk 5 above)
Comment (livecodemichael) [reply Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript]: Haha, reoder. I'm an idiot.

File: Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript (hunk 7 above)
Comment (livecodemichael) [reply Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript]: It's not even close. Where did the n come from?

**Comment by livecodemichael** [comment]
@livecodeali Yeah, I wasn't sure about that. I originally had them as functions since they return a result - it makes sense. But then figured there wasn't really any context elsewhere for doing something similar: The programmer implementing a function that's called by LiveCode felt a bit weird. Handling messages sent by LiveCode felt more typical.

I don't think either method is particularly nice, so I'm happy to go either way. As you say, it's just a single word change (along with docs updates). I don't know if anyone else has an opinion.

File: Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript (hunk 7 above)
Comment (livecodeali) [reply Toolset/palettes/revdatagridlibrary/behaviorsdatagridbuttonbehavior.livecodescript]: Haha, looks like your fingers got so used to typing `constant`. I do it all the type with `local` when I'm trying to lock messages- `locak messages` *all* the time

**Comment by livecodeali** [comment]
@livecodemichael The precedent in my head is `NumberOfRows` and `CacheKeyForRow` in the data view, but then again that's not out there for general use! Ok, let's leave it as `on` then :-)

**Comment by livecodeali** [comment]
@livecode-vulcan review ok 8089bd6

**Comment by livecode-vulcan** [comment]
💙 review by @livecodeali ok 8089bd61a176a96fb181eca4b6b4fa8c32c9be36

**Pull request merged**